### ISSUE TITLE
Refactoring command declarations

### ISSUE THREAD

**Issue by 00vareladavid** [JuliaLang/Pkg.jl#741]
`command_declarations` is becoming unwieldy. Suggestions for a more readable/clear interface?

I have this in mind:
```
[
    :kind => CMD_RM,
    :name => "remove",
    :short_name => "rm",
    :handler => do_rm!,
    :arg_count => 1 => Inf,
    :arg_parser => (parse_pkg, []),
    :completions => complete_rm,
    :option_spec => [
        [:name => "project", :short_name => "p", :api => :mode => PKGMODE_PROJECT],
        [:name => "manifest", :short_name => "m", :api => :mode => PKGMODE_MANIFEST],
    ]
    :description => "remove packages from project or manifest",
]
```

Where `parse_pkg(args::Vector{String})::Vector{PackageSpec}` and `complete_rm(flags::Dict{Symbol,Any})::Vector{String}`. So for a new command you would only need to create a function which returns a list of possibilities, and the general completion code would take care of the rest.

Some fields can be unspecified if a reasonable default is available (e.g. not specifying `option_spec` means that a command does not take any command line flags).

**Comment by 00vareladavid** [comment]
Also, instead of `:handler`, we could have `:api` which would specify an API function that would directly be handed the parsed arguments + command line options. E.g. `:api => API.rm`. This would remove the need for most `do_x!` functions.

**Comment by 00vareladavid** [comment]
Resolved by #818 and #826
